datafusion/proto/tests/cases/roundtrip_physical_plan.rs

```diff
@@ -665,6 +670,133 @@ fn roundtrip_scalar_udf() -> Result<()> {
+        pattern: String,
+    }
+
+    impl MyRegexUdf {
+        fn new(pattern: String) -> Self {
+            Self {
+                signature: Signature::uniform(
+                    1,
+                    vec![DataType::Int32],
+                    Volatility::Immutable,
+                ),
+                pattern,
+            }
+        }
+    }
+
+    /// Implement the ScalarUDFImpl trait for MyRegexUdf
+    impl ScalarUDFImpl for MyRegexUdf {
+        fn as_any(&self) -> &dyn Any {
+            self
+        }
+        fn name(&self) -> &str {
+            "regex_udf"
+        }
+        fn signature(&self) -> &Signature {
+            &self.signature
+        }
+        fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+            if !matches!(args.first(), Some(&DataType::Utf8)) {
+                return plan_err!("regex_udf only accepts Utf8 arguments");
+            }
+            Ok(DataType::Int32)
+        }
+        fn invoke(&self, _args: &[ColumnarValue]) -> Result<ColumnarValue> {
+            unimplemented!()
+        }
+    }
+
+    #[derive(Clone, PartialEq, ::prost::Message)]
+    pub struct MyRegexUdfNode {
+        #[prost(string, tag = "1")]
+        pub pattern: String,
+    }
+
+    #[derive(Debug)]
+    pub struct ScalarUDFExtensionCodec {}
+
+    impl PhysicalExtensionCodec for ScalarUDFExtensionCodec {
+        fn try_decode(
+            &self,
+            _buf: &[u8],
+            _inputs: &[Arc<dyn ExecutionPlan>],
+            _registry: &dyn FunctionRegistry,
+        ) -> Result<Arc<dyn ExecutionPlan>> {
+            not_impl_err!("No extension codec provided")
+        }
+
+        fn try_encode(
+            &self,
+            _node: Arc<dyn ExecutionPlan>,
+            _buf: &mut Vec<u8>,
+        ) -> Result<()> {
+            not_impl_err!("No extension codec provided")
+        }
+
+        fn try_decode_udf(&self, name: &str, buf: &[u8]) -> Result<Arc<ScalarUDF>> {
+            if name == "regex_udf" {
+                let proto = MyRegexUdfNode::decode(buf).map_err(|err| {
+                    DataFusionError::Internal(format!(
+                        "failed to decode regex_udf: {}",
+                        err
+                    ))
+                })?;
+
+                Ok(Arc::new(ScalarUDF::new_from_impl(MyRegexUdf::new(
+                    proto.pattern,
+                ))))
+            } else {
+                not_impl_err!("unrecognized scalar UDF implementation, cannot decode")
+            }
+        }
+
+        fn try_encode_udf(&self, node: &ScalarUDF, buf: &mut Vec<u8>) -> Result<()> {
+            let binding = node.inner();
+            let udf = binding.as_any().downcast_ref::<MyRegexUdf>().unwrap();
+            let proto = MyRegexUdfNode {
+                pattern: udf.pattern.clone(),
+            };
+            proto.encode(buf).map_err(|e| {
+                DataFusionError::Internal(format!("failed to encode udf: {e:?}"))
+            })?;
```

Comment:
nit: this works for the test but as an example I think gives a misleading impressions of how to implement this. We don't want to fail of UDFs which don't need special handling. 

```suggestion
            if let Some(udf) = binding.as_any().downcast_ref::<MyRegexUdf>() {
                let proto = MyRegexUdfNode {
                    pattern: udf.pattern.clone(),
                };
                proto.encode(buf).map_err(|e| {
                    DataFusionError::Internal(format!("failed to encode udf: {e:?}"))
                })?;
            }
```

Reply:
Great, it's a good improvement!